server/punishments.ts

```diff
@@ -31,47 +32,39 @@ const BLACKLIST_DURATION = 365 * 24 * 60 * 60 * 1000; // 1 year
 const USERID_REGEX = /^[a-z0-9]+$/;
 const PUNISH_TRUSTED = false;
 
-/**@type {{[k: string]: number}} */
-const PUNISHMENT_POINT_VALUES = {MUTE: 2, BLACKLIST: 3, BATTLEBAN: 4, ROOMBAN: 4};
+const PUNISHMENT_POINT_VALUES: {[k: string]: number} = {MUTE: 2, BLACKLIST: 3, BATTLEBAN: 4, ROOMBAN: 4};
 const AUTOLOCK_POINT_THRESHOLD = 8;
 
 /**
  * A punishment is an array: [punishType, userid, expireTime, reason]
- * @typedef {[string, string, number, string]} Punishment
- */
-
-/**
- * @typedef {object} PunishmentEntry
- * @property {string[]} ips
- * @property {string[]} userids
- * @property {string} punishType
- * @property {number} expireTime
- * @property {string} reason
- * @property {any[]} rest
  */
+type PunishmentG<T> = [string, T, number, string];
+type Punishment = PunishmentG<ID>;
```

Comment:
`PunishmentG`eneric, but also because I couldn't think of a better name at the time.

I'm specifically typing Punishments#ips as `'#ipban' | ID` because it's the only one that can be a non ID. If every punishment was typed that way you'd need a lot more unnecessary checks for `'#ipban'`

Reply:
Renamed it, I still want `Punishment` to be based off the generic because `PunishmentMap` needs to be typed in terms of the generic.

Reply:
me too, but I can't think of any other solutions that don't involve a ton of asserts / unnecessary checks.